content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+
+A number of challenges face extension users in various configurations, thanks
+to extension file organization in the Postgres core. The common thread among
+them is the need to add extensions without changing the Postgres installation
+itself.
+
+### Packager Testing
+
+On Debian systems, the user account that creates extension packages lacks
+permission to add files to Postgres install. But testing extensions requires
+installing the extension files where Postgres can find them. Furthermore,
+extensions should ideally be built against a clean Postgres install; adding an
+extension in order to run `make installcheck` would pollute it.
+
+[Christoph's patch][destdir] solves these problems by adding a second lookup
+path for extensions and dynamic modules, so that Postgres can load them
+directly from the package build directory.
+
+Alas, the patch isn't ideal, because it simply specifies a prefix and appends
+the full `pg_config` directory paths to it. For example, if `--sharedir`
+outputs `/opt/share` and `extension_destdir` GUC is set to `/tmp/build/myext`,
+the patch will search in `/tmp/build/myext/opt/share`. This approach works for
+the packaging use case, which explicitly uses full paths with a prefix, but
+would be weird for other use cases.
+
+### Docker Immutability
+
+Docker images are immutable. To install persistent extensions in a running
+Docker container, one must create a persistent volume, map it to
+`SHAREDIR/extensions`, and copy over all the core extensions (or muck with
+[symlink magic]). Then do it again for shared object libraries (`PKGLIBDIR`),
+and perhaps also for other `pg_config` directories, like `--bindir`. Once it's
+all set up, one can install a new extension and its files will be distributed
+to the relevant persistent volumes.
+
+This pattern makes upgrades tricky, because the core extensions are mixed in
+with third-party extensions. Worse, the number of directories that must be
+mounted into volumes depends on the features of an extension, increasing
+deployment configuration complexity. It would be preferable to have all the
+files for an extension in one place, rather than scattered across multiple
+persistent volumes.
+
+### Postgres.app Immutability
+
+The macOS [Postgres.app] supports extensions. But installing one into
+`SHAREDIR/extensions` changes the contents of the Postgres.app bundle,
+breaking Apple-required signature validation. The OS will no longer be able to
+validate that the app is legit and refuse to start it.
+
+## Solution
+
+To address these issues, this RFC proposes to change file organization and
+lookup patterns for PostgreSQL extensions.
+
+### Extension Directories
+
+First, when an extension is installed, all of its files will live in a single
+directory named for the extension. The contents include:
+
+*   The Control file that describes extension
+*   Subdirectories for SQL, shared modules, docs, binaries
+
+Subdirectories roughly correspond to the `pg_config --*dir` options:
```

Comment:
I'd avoid bringing any files into extension's directories that are not strictly required for Postgres to load and run that extension. That implies that documentation, man pages, etc, are strictly kept out of this directory. If desired, they can be placed in other system-level directories (e.g. `/usr/share/doc`). This also helps bringing into the directories where Postgres loads stuff the bare minimum amount of files required, which I consider a good thing. Similarly, I'd also keep independent executables out of this directory. Not only they are not to be loaded by Postgres; but they will end up on a path quite unusable for the users --which are the ones expected to run such binaries.

Reply:
I think it's useful to have a complete picture of what all is considered part of an extension, and that includes documentation, license, bill of materials, etc. In other words, it's made for humans and machines.

No reason why one couldn't build layers an an OCI Image that include such files, and use a layer (tag) without them if you want.

Reply:
> I think it's useful to have a complete picture of what all is considered part of an extension, and that includes documentation, license, bill of materials, etc. In other words, it's made for humans and machines.

I agree... but that doesn't mean they all need to be in the same directory. For instance, Debian packages put extension files where they are expected by Postgres (`/usr/lib/postgresql/$major/lib`) and documentation in the usual docs folder (`/usr/share/doc`). I'd do the same.

> No reason why one couldn't build layers an OCI image that include such files, and use a layer (tag) without them if you want.

That's certainly true. However, there's a very practical reason not to abuse layering: certain container implementations have a limit in the number of layers that can be used (e.g. Docker at 127).

Reply:
I have a hard time imagining needing more than 3-4 layers.

Reply:
> I have a hard time imagining needing more than 3-4 layers.

Just imagine you split extensions into two layers (one for the files that PG needs to load to work with the extension; and another one with the docs and other files). Actually you may take this further and also use different layers for each of the extension's dependencies (and that may make sense, as the dependency may not change in a while across extension's versions). But for the sake of the argument let's say 2-3 per extension.

Then if I load, say, 15 extensions, I'm already consuming 30-45 layers.

Add to this the Postgres layer(s) itself --which in many cases is built as several layers, not just one--. And then the "base OS layer", which is normally multiple layers too.

In total you could be at 50-80 layers easily. And while this is what you may think "the maximum under reasonable use cases" the problem is that people will always find a reason to go beyond this "reasonable maximum" and would hit the limit.

All in all, I'd favor "extreme layering", but this unfortunate limitation of some systems like Docker makes it unpractical.

Reply:
> Then if I load, say, 15 extensions, I'm already consuming 30-45 layers.

Oh, you didn't mean per-container? Damn limitations.

Reply:
It's per image. Actually the image doesn't have any limit, it is a limit of some container runtime (like Docker) than when downloading an image with more than 127 layers will refuse to run that container image.

Which may happen if you pack in the image the base OS, Postgres and a few dozen extensions (with every extension having two or more layers).

Yes, you cannot imagine how much I hate this limitation. It's nuts how some limits where imposed by the original Docker designs. That even then made no sense (well, a bit, while `aufs` was used to construct the runtime image, but that time is long gone). There should be always limits (that's defensive programming) but some are arbitrarily and unnecessary low.